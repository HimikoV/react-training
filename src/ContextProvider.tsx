import { createContext, useEffect, useState, useMemo } from "react";
import { mapBasketData } from "./mappers/mapBasketData";
import { getDeliveryOptions } from "./service/getDeliveryOptions";
import { getOrderData } from "./service/getOrderData";
import { getPaymentOptions } from "./service/getPaymentOptions";
import { requestData } from "./service/requestData";
import { basketType } from "./types/basketType";
import { finalizeOptionsType } from "./types/finalizeOptionsType";
import { orderType } from "./types/orderType";
import { productType } from "./types/productType";
import { childrenInterface } from "./interfaces/childrenInterface";
import { basketProcessedType } from "./types/basketProcessedType";
import { contextProviderInterface } from "./interfaces/contextProviderInterface";

export const dataContext = createContext<contextProviderInterface>(
  {} as contextProviderInterface
);

const ContextProvider = ({ children }: childrenInterface) => {
  const [basketData, setBasketData] = useState<basketType[]>(
    JSON.parse(localStorage.getItem("basketData") || "[]")
  );
  const [filter, setFilter] = useState<string[]>([]);
  const [data, setData] = useState<productType[]>([]);
  const [loadLimit, setLoadLimit] = useState<number>(9);
  const [basketProcessedData, setBasketProcessedData] = useState<
    basketProcessedType[]
  >([]);
  const [deliveryOptions, setDeliveryOptions] = useState<finalizeOptionsType[]>(
    []
  );
  const [paymentOptions, setPaymentOptions] = useState<finalizeOptionsType[]>(
    []
  );
  const [open, setOpen] = useState<boolean>(false);
  const [orderData, setOrderData] = useState<orderType[]>([]);

  useEffect(() => {
    const fetchData = async () => {
      const responseData = await requestData();
      setData(responseData);
      const responseDelivery = await getDeliveryOptions();
      setDeliveryOptions(responseDelivery);
      const responsePayment = await getPaymentOptions();
      setPaymentOptions(responsePayment);
      const responseOrderData = await getOrderData();
      setOrderData(responseOrderData);
    };
    fetchData();
  }, []);

  useEffect(() => {
    localStorage.setItem("basketData", JSON.stringify(basketData));
  }, [basketData]);

  useEffect(() => {
    mapBasketData(data, basketData, setBasketProcessedData);
  }, [data]);

  const trueValues = useMemo(
    () => ({
      filter,
      setFilter,
      basketData,
      setBasketData,
      data,
      loadLimit,
      setLoadLimit,
      basketProcessedData,
      setBasketProcessedData,
      deliveryOptions,
      setDeliveryOptions,
      paymentOptions,
      setPaymentOptions,
      open,
      setOpen,
      orderData,
      setOrderData,
    }),
    [
      filter,
      setFilter,
      basketData,
      setBasketData,
      data,
      loadLimit,
      setLoadLimit,
      basketProcessedData,
      setBasketProcessedData,
      deliveryOptions,
      setDeliveryOptions,
      paymentOptions,
      setPaymentOptions,
      open,
      setOpen,
      orderData,
      setOrderData,
    ]
  );

  // const values = {};
  return (
    <dataContext.Provider value={trueValues}>{children}</dataContext.Provider>
  );
};
export default ContextProvider;
