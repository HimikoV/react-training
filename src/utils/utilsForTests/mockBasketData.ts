import { basketProcessedType } from "../../types/basketProcessedType";
import { basketType } from "../../types/basketType";

export const mockBasketData: basketType[] = [
  { id: "item_1", quantity: 4 },
  { id: "item_2", quantity: 5 },
  { id: "item_3", quantity: 1 },
];
export const mockBasketDataResult: basketType[] = [];
export const mockBasketProcessedData: basketProcessedType[] = [];
export const mockBasketProcessedDataResult: basketProcessedType[] = [
  {
    name: "product1",
    price: 15,
    category: "category1",
    id: "item_1",
    imgUrl: "url1",
    quantity: 4,
  },
  {
    name: "product2",
    price: 15,
    category: "category2",
    id: "item_2",
    imgUrl: "url2",
    quantity: 5,
  },
  {
    name: "product3",
    price: 15,
    category: "category3",
    id: "item_3",
    imgUrl: "url2",
    quantity: 1,
  },
];
