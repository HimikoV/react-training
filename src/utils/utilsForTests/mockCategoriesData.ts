import { categoryType } from "../../types/categoryType";

export const mockCategoriesData: categoryType[] = [
  {
    categoryName: "category1",
    categoryId: "category1",
  },
  {
    categoryName: "category2",
    categoryId: "category2",
  },
];
