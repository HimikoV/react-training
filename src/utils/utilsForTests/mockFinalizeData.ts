import { finalizeOptionsType } from "../../types/finalizeOptionsType";

export const mockFinalizeData: finalizeOptionsType[] = [
  {
    name: "DPD Courier",
    imgUrl:
      "https://images.unsplash.com/photo-1601647998802-5c86158216f6?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=755&q=80",
    selected: true,
  },
  {
    name: "DPD Inpost",
    imgUrl:
      "https://images.unsplash.com/photo-1601647998802-5c86158216f6?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=755&q=80",
    selected: false,
  },
];
export const mockFinalizeDataPayment: finalizeOptionsType[] = [
  {
    name: "Cash",
    imgUrl:
      "https://images.unsplash.com/photo-1601647998802-5c86158216f6?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=755&q=80",
    selected: true,
  },
  {
    name: "Card",
    imgUrl:
      "https://images.unsplash.com/photo-1601647998802-5c86158216f6?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=755&q=80",
    selected: false,
  },
];
export const mockFinalizeDataResult: finalizeOptionsType[] = [
  {
    name: "DPD Courier",
    imgUrl:
      "https://images.unsplash.com/photo-1601647998802-5c86158216f6?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=755&q=80",
    selected: true,
    key: "DPD Courier",
  },
  {
    name: "DPD Inpost",
    imgUrl:
      "https://images.unsplash.com/photo-1601647998802-5c86158216f6?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=755&q=80",
    selected: false,
    key: "DPD Inpost",
  },
];
