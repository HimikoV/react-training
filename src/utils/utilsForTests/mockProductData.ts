import { productType } from "../../types/productType";

export const mockProductData: productType[] = [
  {
    name: "product1",
    price: 15,
    category: "category1",
    id: "item1",
    imgUrl: "url1",
  },
  {
    name: "product2",
    price: 15,
    category: "category2",
    id: "item2",
    imgUrl: "url2",
  },
];
export const mockProductDataResult: productType[] = [
  {
    name: "product1",
    price: 15,
    category: "category1",
    id: "item1",
    imgUrl: "url1",
    key: "item1",
  },

  {
    name: "product2",
    price: 15,
    category: "category2",
    id: "item2",
    imgUrl: "url2",
    key: "item2",
  },
];
