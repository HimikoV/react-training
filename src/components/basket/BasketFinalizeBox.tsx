import { Box, Typography, Button } from "@mui/material";
import { useContext, useEffect, useState } from "react";
import { dataContext } from "../../ContextProvider";
import { Link } from "react-router-dom";

const BasketFinalizeBox = () => {
  const { basketProcessedData } = useContext(dataContext);
  const [price, setPrice] = useState<number>(0);

  const addPricesOfBasketItems = () => {
    let prices: number = 0;
    basketProcessedData.forEach((item: any) => {
      prices += item.price * item.quantity;
    });
    setPrice(prices);
  };

  useEffect(() => {
    addPricesOfBasketItems();
  }, [basketProcessedData]);
  return (
    <Box
      sx={{
        display: "flex",
        flexWrap: " wrap",
        color: "#3d405b",
        justifyContent: "flex-start",
        alignItems: "flex-start",
        flexDirection: "column",
        position: "sticky",
        top: 90,
        ml: 5,
        p: 4,
        bgcolor: "#e0ddcb",
        borderRadius: "5%",
      }}
    >
      <Typography sx={{ fontSize: 14 }} color="#3d405b" gutterBottom>
        Products Price: {price}$
      </Typography>
      <Typography sx={{ fontSize: 14 }} color="#3d405b" gutterBottom>
        Delivery Price: {20}$
      </Typography>
      <Typography sx={{ fontSize: 14 }} color="#3d405b" gutterBottom>
        Total Price: {price + 20}$
      </Typography>
      <Button
        component={Link}
        to="/finalize"
        sx={{ bgcolor: "#f4f1de", pb: 0, borderRadius: "0%" }}
      >
        <Typography sx={{ fontSize: 14 }} color="#3d405b" gutterBottom>
          Finalize
        </Typography>
      </Button>
    </Box>
  );
};

export default BasketFinalizeBox;
