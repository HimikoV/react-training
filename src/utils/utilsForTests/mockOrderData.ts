import { deliveryInformationType } from "../../types/deliveryInformationType";
import { orderType } from "../../types/orderType";
import { mockBasketData } from "./mockBasketData";

export const mockDeliveryInformation: deliveryInformationType = {
  name: "asdfasd",
  surname: "sdafas",
  email: "asdf@asdfjha.com",
  phone: "48325235235",
  zipcode: "23-325",
  state: "afdas",
  city: "asdfas",
};

export const mockOrderData: orderType[] = [
  {
    id: 1234,
    items: mockBasketData,
    date: "1/1/2022",
    delivery_information: mockDeliveryInformation,
    delivery_option: "DHL",
    payment_option: "cash",
  },
  {
    id: 1235,
    items: mockBasketData,
    date: "1/1/2022",
    delivery_information: mockDeliveryInformation,
    delivery_option: "DHL",
    payment_option: "cash",
  },
];
