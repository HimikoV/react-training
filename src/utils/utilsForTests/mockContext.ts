import {
  mockBasketData,
  mockBasketProcessedDataResult,
} from "./mockBasketData";
import { mockProductData, mockProductDataResult } from "./mockProductData";
import { mockFinalizeData } from "./mockFinalizeData";
import { mockOrderDataArray } from "./mockOrderData";
import { contextProviderInterface } from "../../interfaces/contextProviderInterface";
import { Firestore } from "firebase/firestore";
import { NULL } from "sass";
import { Auth } from "firebase/auth";

export const mockContext: contextProviderInterface = {
  filter: [],
  setFilter: jest.fn(),
  basketData: mockBasketData,
  setBasketData: jest.fn(),
  data: mockProductData,
  loadLimit: 9,
  setLoadLimit: jest.fn(),
  basketProcessedData: mockBasketProcessedDataResult,
  setBasketProcessedData: jest.fn(),
  deliveryOptions: mockFinalizeData,
  setDeliveryOptions: jest.fn(),
  paymentOptions: mockFinalizeData,
  setPaymentOptions: jest.fn(),
  open: false,
  setOpen: jest.fn(),
  orderData: mockOrderDataArray,
  setOrderData: jest.fn(),
  order: false,
  setOrder: jest.fn(),
  sort: "letter",
  setSort: jest.fn(),
  processedData: mockProductDataResult,
  setProcessedData: jest.fn(),
  orderLoadLimit: 20,
  setOrderLoadLimit: jest.fn(),
  firestore: {} as Firestore,
  auth: {} as Auth,
  user: null,
  setUser: jest.fn(),
};
