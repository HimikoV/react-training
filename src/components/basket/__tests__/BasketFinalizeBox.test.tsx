import { describe, it } from "@jest/globals";
import renderer from "react-test-renderer";
import { dataContext } from "../../../ContextProvider";
import BasketFinalizeBox from "../BasketFinalizeBox";
import { MemoryRouter as Router } from "react-router-dom";
import { render } from "@testing-library/react";
import { screen } from "@testing-library/dom";
import { mockContext } from "../../../utils/utilsForTests/mockContext";

describe("testing BasketFinalizeBox component", () => {
  const wrapper = (linkBool: boolean) => {
    return (
      <Router>
        <dataContext.Provider value={mockContext}>
          <BasketFinalizeBox text="proceed" link={true} />
        </dataContext.Provider>
      </Router>
    );
  };
  it("Renders Correctly without link", async () => {
    const basketFinalizeBox = renderer.create(wrapper(false)).toJSON();
    expect(basketFinalizeBox).toMatchSnapshot();
    // expect(basketFinalizeBox.contains(<Button />)).toBe(true);
    //expect onclick type submit
  });
  it("Renders Correctly with a link", () => {
    const basketFinalizeBox = renderer.create(wrapper(true)).toJSON();
    expect(basketFinalizeBox).toMatchSnapshot();
    //expect is there a button (exist)
    //expect onclick route
  });
  it("check if button exist with link true", () => {
    render(wrapper(true));
    const button = screen.getByTestId("buttonLink");
    expect(button.nodeName).toBe("A");
  });
  it("check if button exist with link false", () => {
    render(wrapper(false));
    const button = screen.getByText("proceed");
    expect(button.nodeName).toBe("SPAN");
  });
});
