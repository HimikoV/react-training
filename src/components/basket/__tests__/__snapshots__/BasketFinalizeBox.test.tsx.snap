// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing BasketFinalizeBox component Renders Correctly with a link 1`] = `
<div
  className="MuiGrid-root css-5jl48k-MuiGrid-root"
>
  <span
    className="MuiTypography-root MuiTypography-caption css-1d9s20a-MuiTypography-root"
  >
    Products Price: 
    150
    $
  </span>
  <span
    className="MuiTypography-root MuiTypography-caption css-1d9s20a-MuiTypography-root"
  >
    Delivery Price: 
    20
    $
  </span>
  <span
    className="MuiTypography-root MuiTypography-caption css-1d9s20a-MuiTypography-root"
  >
    Total Price: 
    170
    $
  </span>
  <a
    className="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-1e6y48t-MuiButtonBase-root-MuiButton-root"
    data-testid="buttonLink"
    href="/finalize"
    onBlur={[Function]}
    onClick={[Function]}
    onContextMenu={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={0}
  >
    <span
      className="MuiTypography-root MuiTypography-caption css-1d9s20a-MuiTypography-root"
    >
      proceed
    </span>
    <span
      className="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
    />
  </a>
</div>
`;

exports[`testing BasketFinalizeBox component Renders Correctly without link 1`] = `
<div
  className="MuiGrid-root css-5jl48k-MuiGrid-root"
>
  <span
    className="MuiTypography-root MuiTypography-caption css-1d9s20a-MuiTypography-root"
  >
    Products Price: 
    0
    $
  </span>
  <span
    className="MuiTypography-root MuiTypography-caption css-1d9s20a-MuiTypography-root"
  >
    Delivery Price: 
    20
    $
  </span>
  <span
    className="MuiTypography-root MuiTypography-caption css-1d9s20a-MuiTypography-root"
  >
    Total Price: 
    20
    $
  </span>
  <a
    className="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-1e6y48t-MuiButtonBase-root-MuiButton-root"
    data-testid="buttonLink"
    href="/finalize"
    onBlur={[Function]}
    onClick={[Function]}
    onContextMenu={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={0}
  >
    <span
      className="MuiTypography-root MuiTypography-caption css-1d9s20a-MuiTypography-root"
    >
      proceed
    </span>
  </a>
</div>
`;
